backfill_corrections/delphiBackfillCorrection/R/model.R

```diff
@@ -262,3 +262,55 @@ generate_filename <- function(indicator, signal,
   )
   return(filename)
 }
+
+#' Get date range of data to use for training models
+#'
+#' Calculate training end date, input data start date, and input
+#' data end date based on user settings.
+#'
+#' Cases:
+#'   1. We are training new models.
+#'   2. We are not training new models and cached models exist.
+#'   3. We are not training new models and cached models don't exist.
+#'
+#' Sometimes we want to allow the user to specify an end date in
+#' params that overrides the automatically-generated end date. This is
+#' only relevant when the user requests to train new models.
+#'
+#' @template params-template
+get_training_date_range <- function(params) {
+  if (params$train_models) {
+    if (params_element_exists_and_valid(params, "training_end_date")) {
+      # Use user-provided end date.
+      training_end_date <- as.Date(params$training_end_date)
+    } else {
+      # Default end date is today.
+      training_end_date <- TODAY
+    }
+  } else {
+    # Get end date from cached model files.
+    # Assumes filename format like `2022-06-28_changehc_covid_state_lambda0.1_count_ca_lag5_tau0.9.model`
+    # where the leading date is the training end date for that model.
+    model_files <- list.files(params$cache_dir, "202[0-9]-[0-9]{2}-[0-9]{2}*.model")
+    if (length(model_files) == 0) {
+      # We know we'll be retraining models today.
+      training_end_date <- TODAY
+    } else {
+      # If only some models are in the cache, they will be used and those
+      # missing will be regenerated as-of the training end date.
+      training_end_date <- max(as.Date(substr(model_files, 1, 10)))
+    }
+  }
+
+  training_start_date <- training_end_date - params$training_days
```

Comment:
what happens if `params$training_days` is different from the `training_days` used to train the stored model? /under what circumstances would we switch to a different value for `training_days`?

Reply:
> what happens if `params$training_days` is different from the `training_days` used to train the stored model?

Yeah, in that case we could end up using an incorrect model from the cache. I think it's unlikely, even with local use testing different training ranges, but it's easy enough to add `training_start_date` into the cached model names.

@jingjtang optimized the current value for `training_days` as part of the model development process, so we'd be unlikely to change it on a short time scale for production runs. I imagine we'd only want to change `training_days` if the signal behavior changed significantly and wasn't being captured well within the current training range.

Reply:
Added this in https://github.com/cmu-delphi/covidcast-indicators/pull/1711/commits/1ecfa81aa1f5f6db02c6e75ad49fa8ecd2c5c932.